frontend/lib/resolution/resolution-queries.cpp

```diff
@@ -1185,6 +1216,114 @@ Type::Genericity getTypeGenericity(Context* context, QualifiedType qt) {
   return getTypeGenericityIgnoring(context, qt, ignore);
 }
 
+/**
+  Written primarily to support multi-decls, though the logic is the same
+  as for single declarations. Sets 'outIsGeneric' with the genericity of the
+  variable; however, if the genericity might be affected by the neighbors
+  of this variable in a MultiDecl, returns 'false'. Thus, calling this function
+  on the neighbors of the variable until it returns 'true' should help determine
+  the genericity in a multi-decl.
+ */
+static bool isVariableDeclWithClearGenericity(Context* context,
+                                              const VarLikeDecl* var,
+                                              bool &outIsGeneric,
+                                              types::QualifiedType* outFormalType) {
+  // fields that are 'type' or 'param' are generic
+  // and we can use the same type/param intent for the type constructor
+  if (var->storageKind() == QualifiedType::TYPE ||
+      var->storageKind() == QualifiedType::PARAM) {
+    if (outFormalType)
+      *outFormalType = QualifiedType(var->storageKind(), AnyType::get(context));
+    outIsGeneric = true;
+    return true;
+  }
+
+  // non-type/param fields with an init expression aren't generic
+  if (var->initExpression() != nullptr) {
+    outIsGeneric = false;
+    return true;
+  }
+
+  // non-type/param fields that have no declared type and no initializer
+  // are generic and these need a type variable for the argument with AnyType.
+  // Except if they are part of a multi-decl, in which case they can
+  // be inheriting their value or type from their neighbor.
+  if (var->typeExpression() == nullptr) {
+    if (outFormalType)
+      *outFormalType = QualifiedType(QualifiedType::TYPE, AnyType::get(context));
+    outIsGeneric = true;
+    return false;
+  }
+
+  // Otherwise, it has a type expression and no init expression. The form
+  // of the type expression determines if we guess it to be generic or
+  // concrete. The only "generic" form is a call with a "(?)" actual.
+
+  if (auto call = var->typeExpression()->toFnCall()) {
+    for (auto actual : call->actuals()) {
+      if (auto ident = actual->toIdentifier()) {
+        if (ident->name() == "?") {
+          outIsGeneric = true;
+          return true;
+        }
+      }
+    }
+  }
+
+  outIsGeneric = false;
+  return true;
+}
+
+bool isFieldSyntacticallyGeneric(Context* context,
+                                 const ID& fieldId,
+                                 types::QualifiedType* formalType) {
+  // compare with AggregateType::fieldIsGeneric
+
+  if (asttags::isVarLikeDecl(parsing::idToTag(context, fieldId))) {
```

Comment:
Since you are about to call `parsing::idToAst` IMO it'd be better to just call that instead of starting with idToTag. I am thinking that we should `CHPL_ASSERT` if fieldId does not refer to a field.